### PR TITLE
Upgrade go-msgpack to v2 2.1.1

### PR DESCRIPTION
**Pull request by swenson**
And set the `time.Time` option to use the go-msgpack-1.1.5-compatible encoding in all the places, since that is the (now previous) version in `go.mod`.

v2 2.1.1 was specifically designed to honor backwards compatibility with 1.1.5 and 0.5.5, and to clean up the code base to be more maintainable. There may performance lost with the 1.1.5 to 2.1.1 migration since the fastpath code was removed, but the increased safety is probably worth it. See
[the release notes for go-msgkack 2.1.0](https://github.com/hashicorp/go-msgpack/releases/tag/v2.1.0) for more details.

I tested this by running this code, and booting up a cluster with a node also running Vault 1.15.0 (before the upgrade). Before I made the changes to set the right `time.Time` option, the previous-version node would throw a bunch of time-decoding errors. After fixing the option, the node came up smoothly, even after changing leadership between them.

This relies on
- [x] https://github.com/hashicorp/raft-boltdb/pull/38
- [x] https://github.com/hashicorp/raft/pull/577

and will need to be updated after they are merged to get the `go.mod` fixes removed.

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -536,3 +537,9 @@ require (
 	sigs.k8s.io/structured-merge-diff/v4 v4.2.3 // indirect
 	sigs.k8s.io/yaml v1.3.0 // indirect
 )
+
+replace github.com/hashicorp/raft => /Users/swenson/projects/raft
+
+replace github.com/hashicorp/raft-boltdb => /Users/swenson/projects/raft-boltdb
+
+replace github.com/hashicorp/raft-boltdb/v2 => /Users/swenson/projects/raft-boltdb/v2
```
Comment (peteski22) [inline go.mod:545]: @swenson I think these may have made it in by accident.

**Comment by banks** [comment]
Thanks for your diligent work on this painful bit of tech debt @swenson ❤️ .

> There may performance lost with the 1.1.5 to 2.1.1 migration since the fastpath code was removed, but the increased safety is probably worth it.

I wonder if we should find a way to quantify this specifically for Vault's usage. I think we only use msgpack for the Raft wire transport (i.e. the thin bit of RPC framing around each Raft RPC with actual log entries already protobuf encoded in the byte slice) and similar log framing for the LogStore (i.e. bit of msgpack boilerplate around each log entries proto bytes to encode the other metadata like index/term/appendedAt time etc.

So it should be relatively easy to write a Benchmark test in our physical/raft package that exercises both network and disk storage and the compare results of several runs with `main` (plus the new benchmark) and this branch. My _guess_ is that it will not be measurably different for each write because the time will be dominated by disk fsyncs and especially on Mac laptops that is super slow - way slower than anything we are likely to measure in encoding a simple object per log committed. I think the quickest way to get around that rather than hacking the code would be to run the benchmark with a directory where a RAMDisk is mounted so the "fsync" is actually just memory updates. This is not "realistic" in that noone should ever run like this in prod, but it acts as a "worst case" overhead - some customers use NVME drives that are much closer to RAM speed than Mac SSD speed for fsyncs so it would be more representative for them!

This is more me musing aloud that requesting you do all this work! If you'd like to I think it would be valuable but we can also probably have someone else try it!

File: go.mod (hunk 1 above)
Comment (swenson) [reply go.mod:545]: It's more of a reminder that I still need to merge those before this is done.

**Comment by swenson** [comment]
Thanks @banks. Time-permitting, I can try to write a benchmark this week.

**Comment by swenson** [comment]
@banks I added a benchmark in 02f613b2832ff73b8ba6f4c6fb4122fce3006e1e

tl;dr there was no performance difference in the benchmark

Though I am suspicious as to why Linux is 10x faster than macOS -- I thought the fsync issue was that macOS didn't do them properly and was supposed to be faster? 🤔

```
goos: darwin
goarch: arm64
pkg: github.com/hashicorp/vault/physical/raft
                   │    a.txt    │            b.txt             │
                   │   sec/op    │   sec/op     vs base         │
RaftWithNetwork-10   58.65m ± 2%   58.62m ± 2%  ~ (p=0.937 n=6)
```

```
goos: linux
goarch: amd64
pkg: github.com/hashicorp/vault/physical/raft
cpu: AMD Ryzen Threadripper 3970X 32-Core Processor
                   │    c.txt    │            d.txt             │
                   │   sec/op    │   sec/op     vs base         │
RaftWithNetwork-64   5.861m ± 1%   5.837m ± 0%  ~ (p=0.240 n=6)
```

File: physical/raft/msgpack.go [hunk 2]
```diff
@@ -9,5 +9,5 @@ package raft
 // on the library, which allows us to pin the version in go.mod.
```
Comment (ncabatoff) [inline physical/raft/msgpack.go:9]: I think we can remove this file now, right?  Or at least update the comment?

File: physical/raft/raft.go [hunk 3]
```diff
@@ -851,6 +852,9 @@ type SetupOpts struct {
 	// RecoveryModeConfig is the configuration for the raft cluster in recovery
 	// mode.
 	RecoveryModeConfig *raft.Configuration
+
+	// overrideMsgpackUseNewTimeFormat is used for testing backwards compatability
+	overrideMsgpackUseNewTimeFormat *bool
```
Comment (ncabatoff) [inline physical/raft/raft.go:857]: What's the motivation here, if Vault has never used the old format?

File: physical/raft/raft_network_test.go [hunk 4]
```diff
@@ -0,0 +1,173 @@
...
+			b.Fatal(err)
+		}
+	}
+}
+
+// TestRaftNetworkClusterWithMultipleTimeEncodingsSet tests that Raft nodes
+// with different msgpack time.Time encodings set will still cluster together.
+// However, with go-msgpack 2.1.0+, the decoder is tolerant of both encodings,
+// so this could only fail if the decoder drastically changes in the future.
+func TestRaftNetworkClusterWithMultipleTimeEncodingsSet(t *testing.T) {
```
Comment (ncabatoff) [inline physical/raft/raft_network_test.go:59]: It seems like this test is verifying that we don't accidentally make a change that results in Vault using the old timestamp format.  But isn't the only way that could happen if we tweak the setting that exists solely for the sake of allowing this test to work?  I'm not sure that it's worth keeping this test on that basis.

File: physical/raft/raft.go (hunk 3 above)
Comment (ncabatoff) [reply physical/raft/raft.go:857]: I guess it's to allow TestRaftNetworkClusterWithMultipleTimeEncodingsSet to exist, but see my comment on that.

File: physical/raft/raft_network_test.go (hunk 4 above)
Comment (swenson) [reply physical/raft/raft_network_test.go:59]: :+1:

**Comment by swenson** [comment]
Thanks everyone. I updated this to remove the extra tests and related code.

**Review by ncabatoff — APPROVED** [review]

**Comment by swenson** [comment]
Thanks!
